### ISSUE TITLE
Implement `float` type

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#262]
Requires:

- `float` type descriptor
- floating point literals
- extend arithmetic operators to float
- extend relational operators to float (these follow IEEE rules)
- extend `==` and `===` to float (these are not the same as IEEE numeric equality)
- extend type cast to float
- io:println on float
- extend tagged pointers to represent float
- float constant evaluation
- BIR instructions for float
- some langlib float functions (exact list TBD)

Note that LLVM hexadecimal float constants can be produced using the langlib float:toBitsInt function.

**Comment by jclark** [comment]
Not sure what `==` and `===` will turn into: maybe need to use LLVM `select` instruction.

**Comment by jclark** [comment]
Also requires extending semtypes to do singleton float types.

**Comment by jclark** [comment]
Another problem is that with the addition of `float`, and integer literal is either of type `int` or `float` depending on the contextually expected type. I will try to deal with this when I have another look at how we handle constants.

**Comment by manuranga** [comment]
~~Until then, I'll restrict the syntax to mandate `f` suffix for non-dotted float literals.~~
A syntax level restriction will not work.

**Comment by manuranga** [comment]
Interesting tokenizer case :`0.e0f()`

**Comment by manuranga** [comment]
Another difficult case `10..toHexString()`.

**Comment by manuranga** [comment]
To handle

**Comment by jclark** [comment]
Please explain the problem. The lexer chooses the longest valid token.

**Comment by jclark** [comment]
That is to say, the longest string that is a token.

**Comment by manuranga** [comment]
Please disregard this, my concerns are baseless since there are no functions in `lang.int` starting with `f` or `e`. I'll use this fact.
I was thinking, if there was,`32.factor()` would be tokenized as `32.f` `actor()`.

**Comment by manuranga** [comment]
When we are parsing a `1.toHexString()` and we see `.`, we still need to see the next char to determine if it's a decimal or float. After setting `t` we need to unget `t` **and** `.` and then return `[DECIMAL_NUMBER, "1"]`. 

(And we are not returning the longest token which is `[DECIMAL_FLOATING_POINT_NUMBER, "1."]`)

**Comment by jclark** [comment]
The way tokenization is supposed to work is to return the longest string that is a valid token. So `1.toHexString()` lexes as the tokens  `1.`   `toHexString`  `(`  `)` which will give a parse error.

**Comment by jclark** [comment]
If you want to do something else, please report it as a spec issue.

**Comment by jclark** [comment]
Another complication to remember is that a type cast expr does conversions between int and float.

**Comment by jclark** [comment]
I think we're going to need to split this into stages: it's too much for one pull request and I think I am going to need to help with some bits.

**Comment by manuranga** [comment]
I'll try to get a minimal substed done and send it as a PR, with just dotted float literals and printing.

Spec mandates that we don't use extra precision but unfortunately there isn't much info on this on llvm side, closest I could find was [this](https://stackoverflow.com/questions/17663780/is-there-a-document-describing-how-clang-handles-excess-floating-point-precision).

This is usually not an issue on modern machines because floating point operations are lowered to xmm instructions. Issues happen if it's an older machine that uses x87 floating point instructions. I have yet to investigate if there are any other ways this edge case can come up (ie xmm registers are exhausted). One easy option would be to disable them altogether, similar to clang flag `-mno-x87`. 

Currently I don't see any need for constrained floating-point intrinsics or strictfp attribute, because (unlike in Java) they only deal with preserving the floating-point exception semantics and non-default rounding. 

For the minimal subset, I'll implement,

* Parse dotted decimal
* Contains check with only uniform float type
* The base representation which is going to be llvm `double`
* Tagged representation and `_bal_float_to_tagged`,  I'll start with always allocating.
* A new type, `RealType` in `llvm.print` corresponding to `LLVMConstReal`. In llvm IR, I can see `double` constants that look like `double 1.000000e+00`. I remember you mentioning hex float literals, do we need those instead?
* `printf` doesn't give me exactly what I need, so use sprintf and massage the result.

**Comment by jclark** [comment]
We need to ensure that we don't get into the situation where we want to close a milestone, but we have only partially implemented floats, i.e. we do not have a correct subset of the language.  There are quite a few tricky things in float, so I am not confident that we will get this done in a single milestone. So we need to stage things very carefully.

So I think we need to do things a bit differently from what you are suggesting. Before you start accepting complete programs involving floats, we should lay the foundations.

1. Design the subset.
2. Tokenize floating point literals (tokenizing but not accepting in the parser keeps everything correct)
3. Extend the types module to support singleton floats (this will be essential for correctness with type narrowing)
4. Add runtime support with runtime tests
5. Design the needed BIR instructions and add to BIR
6. Design the needed extensions to the LLVM module and get them added

All of these can be separate pull requests and none of them negatively affect the correctness of the compiler.

**Comment by jclark** [comment]
When I look at what the LLVM spec says about the double type and about the instructions on it, it doesn't give LLVM any license to use extended precision.  Is a machine that only supports x87 going to support x86_64? I don't think so. I wouldn't spend time on this right now.

**Comment by jclark** [comment]
On tokenization we need to distinguish

- literals that are necessarily of type float - these can be converted to float in the tokenizer
- literals that may be float or decimal depending on the contextually expected type - these should be left as strings in the tokenizer

**Comment by jclark** [comment]
On the runtime:

- I would expect the tagged representation to always allocate on the heap (untagging in the backend should just do a load rather than going via a function call)
- I am surprised to hear you say that `printf` does not do what is needed. How so?

**Comment by jclark** [comment]
On the LLVM module, yes you need to use hex literals.  See the LLVM spec for why https://llvm.org/docs/LangRef.html#simple-constants

**Comment by jclark** [comment]
I think the trickiest thing is going to be https://github.com/ballerina-platform/nballerina/issues/262#issuecomment-887112489 So we should try to figure that out up front.

What I would like to do is to split up the type testing and the conversion at the BIR level. In other words, a type cast that might convert would turn into a regular type cast (that just checks the type) instruction plus a conversion instruction. Please read https://ballerina.io/spec/lang/2021R1/#section_6.20 very carefully.  I think in some cases we may need to split into cast + convert + cast

For example, if we want to do `<1|2>x` where `x` has type `any`, we effectively do:

```
<1|2>convertToInt(<int|float>x)
```

In other words our convertToInt insn requires an operand that is a subtype of `int|float|decimal`.  The backend will compile this differently depending on whether the operand type is a subtype of a single basic type or not.

**Comment by jclark** [comment]
On the LLVM side, as well as `fadd` etc, we are going to need conversion operators. For int to float I think we can use `sitofp`. For float to int  we cannot use fptosi because we need an error if the conversion cannot be done. See https://ballerina.io/spec/lang/2021R1/#section_5.8.4 Maybe there's an intrinsic we can use

Note that conversion has to be handled correctly in the milestone that adds `float` support.

**Comment by jclark** [comment]
I suggest we create some issues for separable parts of this. This will give us somewhere to have more focused discussions on each area.

**Comment by jclark** [comment]
I take charge of getting the front end complete, and then create specific issues for everything in the backend.

**Comment by jclark** [comment]
Things that don't yet have their own issue:

- extend arithmetic operators to float
- extend relational operators to float (these follow IEEE rules)
- add BIR instructions for conversion to float and to int
- extend implementation of BIR type test insn to float
- extend implementation of BIR type cast insn to float
- some langlib float functions (exact list TBD)

**Comment by jclark** [comment]
I think we now have issues for everything that needs doing in the backend.

**Comment by jclark** [comment]
This is done apart from #323.
